File: src/main/core/itunesdb.ts
/**
 * iTunesDB 只读解析 —— 仅用于**取回标题等文本元数据**。
 *
 * 为什么需要它：实测本机 13 个 MP3 中有 4 个完全没有 ID3 标签、其余也没有标题字段，
 * 歌名只存在于 iTunesDB 里。若不做这层回退，界面上只能显示 `SBJT` 这类随机文件名
 * ——正是本工具要解决的问题。
 *
 * 重要：shuffle 播放**不读**这个文件，因此本模块**只读、永不写入**。
 */
import * as fs from 'fs';
import * as path from 'path';

export interface DbEntry {
  title: string;
  album: string;
  artist: string;
  genre: string;
}

/** mhod 类型编号 → 字段名 */
const MHOD_TITLE = 1;
const MHOD_PATH = 2;
const MHOD_ALBUM = 3;
const MHOD_ARTIST = 4;
const MHOD_GENRE = 5;

/**
 * 读取 iTunesDB，返回 `路径 → 元数据` 的映射。
 *
 * 键为「去掉前导 `/` 的 iPod 相对路径」（如 `iPod_Control/Music/F00/SBJT.mp3`），
 * 以便与 iTunesSD 中的 `filename` 字段对齐。
 *
 * 任何结构性异常都只返回空映射，不抛错 —— 元数据回退是「锦上添花」，
 * 不能因为 iTunesDB 损坏就阻断整个流程。
 */
export function readItunesDb(ipodRoot: string): Map<string, DbEntry> {
  const out = new Map<string, DbEntry>();
  const file = path.join(ipodRoot, 'iPod_Control', 'iTunes', 'iTunesDB');
  let db: Buffer;
  try {
    db = fs.readFileSync(file);
  } catch {
    return out;
  }

  try {
    const i = db.indexOf('mhlt');
    if (i < 0) return out;
    const hl = db.readUInt32LE(i + 4);
    const n = db.readUInt32LE(i + 8);

    let q = i + hl;
    for (let t = 0; t < n; t++) {
      if (q + 16 > db.length || db.toString('latin1', q, q + 4) !== 'mhit') break;
      const mhitHeaderLen = db.readUInt32LE(q + 4);
      const mhitTotalLen = db.readUInt32LE(q + 8);
      const nHod = db.readUInt32LE(q + 12);

      const fields = new Map<number, string>();
      let c = q + mhitHeaderLen;
      for (let h = 0; h < nHod; h++) {
        if (c + 16 > db.length || db.toString('latin1', c, c + 4) !== 'mhod') break;
        const hodHeaderLen = db.readUInt32LE(c + 4);
        const hodTotalLen = db.readUInt32LE(c + 8);
        if (hodTotalLen <= 0) break;
        const type = db.readUInt32LE(c + 12);
        const strLenAt = c + hodHeaderLen + 4;
        const dataAt = c + hodHeaderLen + 16;
        if (strLenAt + 4 <= db.length && dataAt <= db.length) {
          const strLen = db.readUInt32LE(strLenAt);
          const end = Math.min(dataAt + strLen, db.length);
          if (end > dataAt) {
            fields.set(type, db.toString('utf16le', dataAt, end));
          }
        }
        c += hodTotalLen;
      }

      // iTunesDB 的路径用 ':' 分隔（iTunesSD 用 '/'）
      const rawPath = (fields.get(MHOD_PATH) ?? '').replace(/:/g, '/');
      if (rawPath) {
        out.set(rawPath.replace(/^\/+/, ''), {
          title: fields.get(MHOD_TITLE) ?? '',
          album: fields.get(MHOD_ALBUM) ?? '',
          artist: fields.get(MHOD_ARTIST) ?? '',
          genre: fields.get(MHOD_GENRE) ?? '',
        });
      }
      if (mhitTotalLen <= 0) break;
      q += mhitTotalLen;
    }
  } catch {
    // 结构损坏：返回已解析到的部分
  }
  return out;
}

File: tauri-app/src-tauri/src/itunesdb.rs
use std::collections::HashMap;
use std::fs;
use std::path::Path;

#[derive(Debug, Clone)]
#[allow(dead_code)]
pub struct DbEntry {
    pub title: String,
    pub album: String,
    pub artist: String,
    pub genre: String,
}

const MHOD_TITLE: u32 = 1;
const MHOD_PATH: u32 = 2;
const MHOD_ALBUM: u32 = 3;
const MHOD_ARTIST: u32 = 4;
const MHOD_GENRE: u32 = 5;

fn rd_u32_le(data: &[u8], off: usize) -> u32 {
    u32::from_le_bytes([data[off], data[off+1], data[off+2], data[off+3]])
}

pub fn read_itunes_db(ipod_root: &str) -> HashMap<String, DbEntry> {
    let mut out = HashMap::new();
    let file = Path::new(ipod_root).join("iPod_Control").join("iTunes").join("iTunesDB");
    let db = match fs::read(&file) {
        Ok(d) => d,
        Err(_) => return out,
    };

    let _result = (|| -> Option<()> {
        // Find mhlt
        let i = db.windows(4).position(|w| w == b"mhlt")?;
        if i + 12 > db.len() { return None; }
        let hl = rd_u32_le(&db, i + 4) as usize;
        let n = rd_u32_le(&db, i + 8) as usize;

        let mut q = i + hl;
        for _t in 0..n {
            if q + 16 > db.len() { break; }
            if &db[q..q+4] != b"mhit" { break; }
            let mhit_header_len = rd_u32_le(&db, q + 4) as usize;
            let mhit_total_len = rd_u32_le(&db, q + 8) as usize;
            let n_hod = rd_u32_le(&db, q + 12) as usize;

            let mut fields: HashMap<u32, String> = HashMap::new();
            let mut c = q + mhit_header_len;
            for _h in 0..n_hod {
                if c + 16 > db.len() || &db[c..c+4] != b"mhod" { break; }
                let hod_header_len = rd_u32_le(&db, c + 4) as usize;
                let hod_total_len = rd_u32_le(&db, c + 8) as usize;
                if hod_total_len == 0 { break; }
                let mhod_type = rd_u32_le(&db, c + 12);
                let str_len_at = c + hod_header_len + 4;
                let data_at = c + hod_header_len + 16;
                if str_len_at + 4 <= db.len() && data_at <= db.len() {
                    let str_len = rd_u32_le(&db, str_len_at) as usize;
                    let end = (data_at + str_len).min(db.len());
                    if end > data_at {
                        // UTF-16LE decode
                        let raw = &db[data_at..end];
                        let u16s: Vec<u16> = raw.chunks_exact(2)
                            .map(|c| u16::from_le_bytes([c[0], c[1]]))
                            .collect();
                        let s = String::from_utf16_lossy(&u16s);
                        fields.insert(mhod_type, s);
                    }
                }
                c += hod_total_len;
            }

            let raw_path = fields.get(&MHOD_PATH).cloned().unwrap_or_default()
                .replace(':', "/");
            if !raw_path.is_empty() {
                let key = raw_path.trim_start_matches('/').to_string();
                out.insert(key, DbEntry {
                    title: fields.get(&MHOD_TITLE).cloned().unwrap_or_default(),
                    album: fields.get(&MHOD_ALBUM).cloned().unwrap_or_default(),
                    artist: fields.get(&MHOD_ARTIST).cloned().unwrap_or_default(),
                    genre: fields.get(&MHOD_GENRE).cloned().unwrap_or_default(),
                });
            }

            if mhit_total_len == 0 { break; }
            q += mhit_total_len;
        }
        Some(())
    })();

    out
}